### ISSUE TITLE
createRigidbody nur im Konstruktor?

### ISSUE THREAD

**Issue by JirkaDellOro** [JirkaDellOro/FUDGE#258]
Bislang habe ich den Aufruf von createRigidbody nur im Konstruktor von ComponentRigidbody entdeckt. Da der Konstruktor durch ist, bis die Komponente ihren Container erhält, könnte man die Zeile

```typescript
     let tmpTransform: Matrix4x4 = _mtxTransform == null ? super.getContainer() != null ? super.getContainer().mtxWorld : Matrix4x4.IDENTITY() : _mtxTransform; //Get transform informations from the world, since physics does not care about hierarchy
```
vereinfachen auf
```typescript
     let tmpTransform: Matrix4x4 = _mtxTransform || Matrix4x4.IDENTITY();
```

oder war noch geplant, dass die Rigidbodies zur Laufzeit rekreiert werden können?

**Comment by Dev-MarkoF** [comment]
Alte Überreste, müsste entfernt werden können.

Oimo, wie die meisten Physik Engines hat nicht komplett bedacht, dass in Games Rigidbodies zur Laufzeit skaliert werden können (bzw. Collider), daher hat Oimo diesen Fehler drin gehabt, durch den bei einer Skalierung, der Body rekreiert hätte werden müssen. Ich hatte das bei Oimo angemerkt und mit Hilfe eines anderen Oimo Nutzers gefixt und konnte ohne Rekreation des Körpers auskommen.

Das ist bis Heute nicht gefixt im Original und auch meine Anfrage für die interne Verarbeitung von Triggern nicht, weshalb Trigger ja über aabb Tests umgesetzt sind, anstatt 100% akkurate Überlappung der Collider.

**Comment by JirkaDellOro** [comment]
> Ich hatte das bei Oimo angemerkt und mit Hilfe eines anderen Oimo Nutzers gefixt und konnte ohne Rekreation des Körpers auskommen.   

Wenn ich mich recht erinnere, haben wir also derzeit in FUDGE eine speziell angepasste Version von OIMO, korrekt?

> weshalb Trigger ja über aabb Tests umgesetzt sind, anstatt 100% akkurate Überlappung der Collider.

OK, das erklärt dann wohl #256. Wieviel Act wäre es, dass akkurat zu machen? Offensichtlich ist ja eigentlich schon alles da, denn die Nicht-Trigger sind akkurat, korrekt?

>  da sich z.B. ein Heightmap Collider "relativ" gut drin umsetzen  

Wenn Du dazu ein paar Zeilen schreibst, kann ich gut eine Diskussion daraus machen und jemand anderes kann es implementieren.

**Comment by Dev-MarkoF** [comment]
> Wenn ich mich recht erinnere, haben wir also derzeit in FUDGE eine speziell angepasste Version von OIMO, korrekt?
Genau, weil Skalierungen von Bodies zur Laufzeit im Original fehlerhaft sind.

> OK, das erklärt dann wohl #256. Wieviel Act wäre es, dass akkurat zu machen? Offensichtlich ist ja eigentlich schon alles da, denn die Nicht-Trigger sind akkurat, korrekt?
Es wäre ein großer Akt und nur in Oimo umsetzbar, auf Fudge Seite gäbe es da nichts zu tun. Trigger werden in Realphysik nicht betrachtet, daher sind die in Engines häufig garnicht, oder halb implementiert. Oimo hat sie gar nicht, und Ammo hat die Callbacks dafür rausgenommen, weswegen BabylonJS z.B. eine ältere Version von Ammo nutzte (Zeitpunkt der Thesis, kann inzwischen wieder drin sein). 
Was es dafür benötigt, wäre in Oimo eine Flag für den RB, bei der in der normalen Kollisionsberechnung, die Kollision normal festgestellt wird, aber das Resultat für die wirkenden Kräfte des Frames verworfen werden. (Keine genauen Zeilen parat, ich hatte mir das mal angeschaut, wären 2-3 Klassen die in Oimo vermutlich angepasst werden müssten)

> da sich z.B. ein Heightmap Collider "relativ" gut drin umsetzen
Auch hier muss nur Oimo angepasst werden, in Fudge wäre das nur ein neuer auswählbarer Collider.
Oimo besitzt dankbarerweise Basiscollider und die erbenden Collider Box, etc. In Physik Engines muss ein Solver für jeden Collider vs jeden Collider existieren (sonst wird halt nicht berechnet, geht aber auch). So gibt es z.B. in Engines wie Cannon die Möglichkeit von Concave Collider (Trimesh), die aber dann z.B. nur mit Sphären kollidieren (weil das die einfachste Form der Kollision ist, beschrieben in der Thesis)
In Oimo müsste, also nur die Definition eines Heightmap Colliders hinzugefügt werden und die Solver pro anderem Collider. 
Kann man angepasst wahrscheinlich aus anderen Engines extrahieren. Aber Oimo ist in Haxe geschrieben, da müsste man sich insgesamt mit dem building process zu js kurz befassen.

Ich habs im Hinterkopf. und schaue vlt. Zeitweise mal rein, weils mich schon reizt, weiß nicht ob das jemand anders so einfach machen kann. Oimo ist sehr strukturiert aufgebaut, das ist wirklich Topqualität im Vergleich zu z.B. Ammo, aber es sind natürlich auch Annahmen getroffen worden, die ggf. Anpassungen erschweren und es ist nur ein "aktiver" Entwickler.

**Comment by Dev-MarkoF** [comment]
Kleine Notiz: Der Entwickler scheint wieder an Oimo zu arbeiten und ist interessiert daran, trigger ordentlich zu implementieren. Ich behalte das im Auge und update ggf. die Version und Anbindung.

**Comment by JirkaDellOro** [comment]
Top. Bist Du mit ihm in Kontakt? Vielleicht können wir ihm an manchen Stellen etwas unter die Arme greifen...

**Comment by Dev-MarkoF** [comment]
Nicht direkt im Kontakt, ich habe nur inzwischen auf mein Trigger Issue eine Antwort erhalten und es gab wieder ein paar Commits von ihm in der letzten Zeit, außerdem Antworten auf andere Issues, daher gehe ich von mehr Aktivität aus. 
Vor allem im Bereich Trigger/Ghosts scheint etwas geplant zu sein, weil ich nicht der einzige bin, der das fordert. 
Diese webbasierten Physikengines sind alle sehr eigen, da tut sich viel und dann lange Zeit gar nichts, wie gesagt muss ich im Auge behalten.

**Comment by JirkaDellOro** [comment]
Wenn Du den Editor mit den Joints ein wenig geprüft hast und wir eine hübsche Test-Szene dafür gebaut haben, solltest Du den Entwickler einmal darauf aufmerksam machen. Ich kann mir vorstellen, dass es ihn interessiert zu sehen, wie man seine Objekte interaktiv parametriert.
